### PR TITLE
pass encoding=utf-8 keyword argument to flask.json.dumps

### PR DESCRIPTION
**Pull request by LeenaBhegade**
Issue #395 
Passing keyword argument "encoding=utf-8" to flask.json.dumps to make sure responses are encoded using utf-8.

### CODE REVIEW

**Comment by hjacobs** [comment]
Thanks for the PR! Currently all the Travis CI tests fail.. (haven't checked the root cause)

**Comment by rafaelcaricio** [comment]
For this PR there is an actual problem. Since we are encoding now with `utf-8` the returned valus are different and tests are failing. Example:

```
    def test_path_parameter_somefloat(simple_app):
        app_client = simple_app.app.test_client()
        resp = app_client.get('/v1.0/test-float-path/123.45')  # type: flask.Response
>       assert resp.data.decode() == '"float"\n'
E       assert 'b\'"float"\'\n' == '"float"\n'
E         - b'"float"'
E         ? --       -
E         + "float"
```

The endpoints are returning `b'"float"'` instead of `"float"` and it is a valid JSON. Then all `json.loads` in the Connexion response are failing.

**Comment by LeenaBhegade** [comment]
Hi @rafaelcaricio 

Thanks for your comment.
In my case for client side testing swagger has generated code for me  - resp.data.decode('utf-8'). 
Python 2.7 will by default use ascii for resp.data.decode()

From web browsers when I trigger the request the responses are decoded properly in utf-8 without any issues.

If we pass 'utf-8' parameter to the decoding method it should work I think. 

If you don't want that change then I can alternatively pass the encoding=utf-8 parameter conditionally when Flask config parameter app.app.config['JSON_AS_ASCII']=False is true.

Please let me know what you think is the best option ?

**Comment by rafaelcaricio** [comment]
@LeenaBhegade I think the best option is to update the tests then to decode properly with 'utf-8'. Thank you for working on this.

**Comment by rafaelcaricio** [comment]
BTW, please rebase with the latest changes from master.

**Comment by LeenaBhegade** [comment]
Hi @rafaelcaricio 

I am happy to make changes to the tests. 

Thanks,
Leena

**Comment by LeenaBhegade** [comment]
hi @rafaelcaricio,

I have added the "utf-8" parameter in the decoding function in all the tests. This has fixed all the test cases in python 2.x tests.

 But python 3.x  test cases are failing due to two errors
1) Assertion error assert 'b\'"DummyClass"\'\n' == '"DummyClass"\n'
2) json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

The decode function in Python3.x somehow is not stripping the prefix 'b' from the response. This is failing the 1) assertion and 2) json.loads (as it expects a bracket and not a character at the beginning)

A prefix of 'b' or 'B' is ignored in Python 2. Python 3 clearly represent byte string and they are still valid strings.

We could either not perform any encoding in python 3.x as utf-8 encoding is well handled in Python3 internally or strip the prefix b's off before assertion and json.loads.

I will continue to check further in details if it is some other reason for the problem.

**Comment by rafaelcaricio** [comment]
@LeenaBhegade The way to go is to do a conditional for Python2.7 and Python3.x. Splitting the "b" from the string representation is not an acceptable solution. You can use six package for that.

**Comment by LeenaBhegade** [comment]
hi @rafaelcaricio 

Thanks for your comment. I will implement the changes soon.

Thanks,
Leena

**Comment by rafaelcaricio** [comment]
@LeenaBhegade Thank you for taking care of this!

**Comment by LeenaBhegade** [comment]
@rafaelcaricio  No worrries. I am glad I could make some contribution towards it.

File: connexion/decorators/produces.py [hunk 1]
```diff
@@ -76,7 +77,12 @@ def dumps(data):
         """ Central point where JSON serialization happens inside
         Connexion.
         """
-        return "{}\n".format(json.dumps(data, indent=2))
+        if six.PY2 == True:
```
Comment (rafaelcaricio) [inline connexion/decorators/produces.py]: `if six.PY2:` should do.

File: connexion/decorators/produces.py [hunk 2]
```diff
@@ -76,7 +77,12 @@ def dumps(data):
...
+            encoding = "utf-8"
+        else:
+            encoding = None
+
+        return "{}\n".format(json.dumps(data, indent=2, encoding=encoding))
```
Comment (rafaelcaricio) [inline connexion/decorators/produces.py]: Maybe better to do:
```python
    json_content = json.dumps(data, indent=2, encoding='utf-8')
else:
    json_content = json.dumps(data, indent=2)

return "".format(json_content)
```

Comment (LeenaBhegade) [inline connexion/decorators/produces.py]: Thanks for your comments. Will do.

Comment (LeenaBhegade) [inline connexion/decorators/produces.py]: @rafaelcaricio,

you have suggested return "".format(json_content) 
but why are we not doing  return "{}\n".format(json_content) as it was being returned originally ?

Comment (hjacobs) [inline connexion/decorators/produces.py]: @LeenaBhegade I assume it was a typo and @rafaelcaricio meant "{}\n".format(...)

**Comment by rafaelcaricio** [comment]
My mistake. I would prefer to leave it exactly as it was before.

**Comment by rafaelcaricio** [comment]
@LeenaBhegade You can add as many test cases as you think is necessary.

**Comment by LeenaBhegade** [comment]
Sorry I have been pushing the code before testing the test cases. Somehow can't get the test cases running locally. Continuously getting module import errors.

**Comment by rafaelcaricio** [comment]
@LeenaBhegade It's ok, you can squash the commits later.

**Comment by LeenaBhegade** [comment]
@rafaelcaricio 

All the tests are passing except for Python 3.5 version  --- >ERROR: InvocationError: '/home/travis/build/zalando/connexion/.tox/flake8/bin/python setup.py flake8'

If the above error is not related to this change and you are happy with the test and changes, please could you approve the changes.

File: connexion/decorators/produces.py (hunk 1 above)
Comment (rafaelcaricio) [inline connexion/decorators/produces.py]: the `== True` is not necessary here.

**Comment by rafaelcaricio** [comment]
@LeenaBhegade Thank you for working on this. I would like to kindly ask you to squash your commits. Please find out more about it here: https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/

**Comment by LeenaBhegade** [comment]
@rafaelcaricio Done. Please approve whenever you get a chance.

**Comment by rafaelcaricio** [comment]
@LeenaBhegade I would like to point out one of my code review requests was ignored: https://github.com/zalando/connexion/pull/397/files#r100395476

Please change that so we can get this merged.

**Comment by LeenaBhegade** [comment]
@rafaelcaricio  
Sorry about missing that out. I have made the changes.
Thanks
Leena

**Comment by rafaelcaricio** [comment]
Please apply this last change:
```patch
--- /home/travis/build/zalando/connexion/connexion/decorators/produces.py:before	2017-02-12 12:47:19.333433
+++ /home/travis/build/zalando/connexion/connexion/decorators/produces.py:after	2017-02-12 12:50:03.352099
@@ -2,9 +2,9 @@
 import datetime
 import functools
 import logging
-import six
 
 import flask
+import six
 from flask import json
 
 from .decorator import BaseDecorator
```

Basically it is just to move the line `import six` line to after `import logging`.

I am sorry to ask so many changes not directly related to your primary objective. But we need to keep the code quality of patches. I know it may be frustrating, but I hope you understand.

All those checks you could have executed in your local machine by running the command `tox` before pushing to the remote branch. Of course you would need to install all python versions we need to run the tests and checks.

**Comment by LeenaBhegade** [comment]
Np.I apologise for missing out these little things. I should have taken care.

**Comment by rafaelcaricio** [comment]
👍

**Comment by rafaelcaricio** [comment]
@LeenaBhegade Thank you for working on this. PR merged! :1st_place_medal:

**Comment by LeenaBhegade** [comment]
Thanks for letting me to do the work!

**Comment by LeenaBhegade** [comment]
Hi @rafaelcaricio 

Just wanted to know when I would be able to install the new version of  connexion via pip ?

Thanks,
Leena

**Comment by rafaelcaricio** [comment]
@LeenaBhegade I will be releasing a new version today. I will let you know.

**Comment by rafaelcaricio** [comment]
@LeenaBhegade Sorry the delay. A new version of Connexion with your contribution has been just released https://pypi.python.org/pypi/connexion version `1.1.5`. Thank you again! 👍

**Pull request merged**